core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala

```diff
@@ -0,0 +1,157 @@
+}
+import org.apache.spark.{ SparkEnv, TaskContext }
+import org.apache.spark.eventhubs.{ EventHubsConf, NameAndPartition, SequenceNumber }
+import org.apache.spark.internal.Logging
+
+import scala.collection.mutable.ListBuffer
+
+private[spark] trait CachedReceiver {
+  def receive(ehConf: EventHubsConf,
+              nAndP: NameAndPartition,
+              requestSeqNo: SequenceNumber,
+              batchSize: Int): EventData
+}
+
+private class CachedEventHubsReceiver(ehConf: EventHubsConf, nAndP: NameAndPartition)
+    extends Logging {
+
+  import org.apache.spark.eventhubs._
+
+  val prefetchCounts = new FixedList[Int](5)
+
+  private var _client: EventHubClient = _
+  private def client: EventHubClient = {
+    if (_client == null) {
+      _client = ClientConnectionPool.borrowClient(ehConf)
+    }
+    _client
+  }
+
+  def createReceiver(requestSeqNo: SequenceNumber): Unit = {
+    logInfo(s"creating receiver for Event Hub ${nAndP.ehName} on partition ${nAndP.partitionId}")
+    val consumerGroup = ehConf.consumerGroup.getOrElse(DefaultConsumerGroup)
+    val receiverOptions = new ReceiverOptions
+    receiverOptions.setReceiverRuntimeMetricEnabled(false)
+    receiverOptions.setIdentifier(s"${SparkEnv.get.executorId}-${TaskContext.get.taskAttemptId}")
+    _receiver = client.createReceiverSync(consumerGroup,
+                                          nAndP.partitionId.toString,
+                                          EventPosition.fromSequenceNumber(requestSeqNo).convert,
+                                          receiverOptions)
+    _receiver.setPrefetchCount(DefaultPrefetchCount)
+  }
+
+  private[this] var _receiver: PartitionReceiver = _
+  private def receiver: PartitionReceiver = {
+    if (_receiver == null) {
+      throw new IllegalStateException(s"No receiver for $nAndP")
+    }
+    _receiver
+  }
+
+  def errWrongSeqNo(requestSeqNo: SequenceNumber, receivedSeqNo: SequenceNumber): String =
+    s"requestSeqNo $requestSeqNo does not match the received sequence number $receivedSeqNo"
+
+  def receive(requestSeqNo: SequenceNumber, batchSize: Int): EventData = {
+    @volatile var event: EventData = null
+    @volatile var i: java.lang.Iterable[EventData] = null
+    while (i == null) {
+      i = receiver.receiveSync(1)
+    }
+    event = i.iterator.next
+
+    if (requestSeqNo != event.getSystemProperties.getSequenceNumber) {
+      logWarning(
+        s"$requestSeqNo did not match ${event.getSystemProperties.getSequenceNumber}." +
+          s"Recreating receiver for $nAndP")
+
+      createReceiver(requestSeqNo)
+
+      while (i == null) {
+        i = receiver.receiveSync(1)
+      }
+      event = i.iterator.next
+      assert(requestSeqNo == event.getSystemProperties.getSequenceNumber,
+             errWrongSeqNo(requestSeqNo, event.getSystemProperties.getSequenceNumber))
+    }
+
+    logAndUpdatePrefetch(batchSize)
+    event
+  }
+
+  private def logAndUpdatePrefetch(batchSize: Int): Unit = {
```

Comment:
lets updatePrefetch to latest BatchSize to keep the code simple and behavior predictable.